### ISSUE TITLE
{35} Future implementation of local commands (against AQL)

### ISSUE THREAD

**Issue by benjamin-actyx** [Actyx/Actyx#90]
### Current impl:
1. command runs against current state of a Fish
2. command execution creates 0-n events
3. command pipeline filters the created events by "does it match the Fish’s query?"
4. selects last event (highest lamport) that matches the query
5. execution of next command is stalled until that event has arrived in the Fish
____
Step 3 is easy to do for tag queries, but with AQL… we don’t want to parse AQL in the client at all.

So there are multiple options for the future:

### Don’t support running commands against AQL 

…only against simple tag expressions.

### Extend AQL with a "command" feature / bi-directional communication

Maybe in the direction of https://github.com/Actyx/Cosmos/pull/5637.
To still allow arbitrary client code to execute, this would require some sort of callback-channel into the ongoing AQL query.
Basically, the query’s ID (already part of `subscribe_monotonic`) is used to refer to it when emitting the events, and this… leads to some desired effect. E.g. it could trigger the offsets reporting explained in the next point.

### Stream all offset-updates into the Fish / Local Twin

Meaning: Report to the Fish when offsets advance, even if they do not match its query. See https://github.com/Actyx/Cosmos/issues/6869
Obviously the updates could be compact, including only the changed sources

### When emitting events, attach the Fish’s query, store reports matches

Options ordered by performance descending:
- use session id to refer to the ongoing query
- use the raw string to find the cached version of the query
- parse the complete query again every time
